### DISCUSSION TITLE
Ability to cancel I/O bound operations

### DISCUSSION THREAD

**Discussion by codercms** [swow/swow#122] [General]
Hello! Do you have any plans to implement ability to cancel I / O operations like GoLang does?

https://golang.org/pkg/context/

Package context defines the Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes.

I think it should be a core ability. It can be useful for services graceful shutdown and can be more consistent with I / O operations (like context.WithTimeout () can be used to cancel any I / O operation by timeout only in one way instead of many setTimeout APIs on Socket and etc.)

**Comment by codercms** [comment]
// Created by mistake (Golang can't cancel reading from socket , only timeout values can be set)
Context can be implemented at user level

**Comment by twose** [comment]
Yeah, Golang cannot really cancel IO operations, all you can do is ignore the IO operation.
But, Swow support cancels all blocking operations, just call `$coroutine->resume()` then you can force resume the blocking coroutine and you would get `Errno\ECANCELED` error, you also can call `$coroutine->throw()` to throw any exceptions to cancel it and catch it then do something that you like, or you can call `$coroutine->kill()` to kill the coroutine, just like you kill the system process.
None of the above actions will have any side effects, this shows the power of Swow.

**Comment by codercms** [comment]
@twose thanks for your answer!
